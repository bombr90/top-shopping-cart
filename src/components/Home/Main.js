import Content from './Content'

const Main = () => {
  
  //////////////////////////////////////
  ///Return (render) ///////////////////
  //////////////////////////////////////

  return (
    <div id="mainHome">
      <Content />
    </div>
  );
}

export default Main;
