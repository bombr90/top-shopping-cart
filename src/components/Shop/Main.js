import React, {useState} from "react";
import uniqid from "uniqid";
import products from './../../data/data'
import Cart from './Cart';
import Catalogue from "./Catalogue";
import {shuffle} from './../util'

// Create local copy of JSON data for store maintained while on the shopping page
  
const productsList = (() => {
  const productsCopy = products.map((product) => {
    product.uid = uniqid();
    return product;
  });
  return shuffle(productsCopy)
})();

const Main = () => {
  const [shoppingCart, setShoppingCart] = useState([]);

  const [cartState, setCartState] = useState({
    display: false,
  });

  const [shoppingCatalogue, setShoppingCatalogue] = useState({
    products: productsList,
  });

  //////////////////////////////////////
  ///Shopping Cart Callbacks////////////
  //////////////////////////////////////

  const handleRemoveClick = (event) => {
    const uid = event.target.parentElement.id;
    setShoppingCart((prev) => {
      return prev.filter((el) => el.uid !== uid);
    });
  };

  const handleCheckCart = () => {
    alert("Checkout functionality not coded.")
  };

  const handleCloseCart = () => setCartState({ display: false });

  const handleEmptyCart = () => {
    return window.confirm("Are you sure you want to empty your cart?")
      ? setShoppingCart([])
      : false;
  };

  const handleIncreaseQuantity = (event) => {
    const uid = event.target.parentElement.parentElement.id;
    setShoppingCart((prev) => {
      return prev.map((el) =>
        el.uid === uid ? { ...el, quantity: Math.min(el.quantity++, 99) } : el
      );
    });
  };

  const handleDecreaseQuantity = (event) => {
    const uid = event.target.parentElement.parentElement.id;
    setShoppingCart((prev) => {
      return prev.map((el) =>
        el.uid === uid ? { ...el, quantity: Math.max(el.quantity--, 1) } : el
      );
    });
  };

  const handleQuantityChange = (event) => {
    const value = event.target.value;
    if (value < 0 || value > 99) return;
    const uid = event.target.parentElement.parentElement.id;
    setShoppingCart(
      shoppingCart.map((el) =>
        el.uid === uid
          ? { ...el, quantity: Math.max(1, value, Math.min(value, 99)) }
          : el
      )
    );
  };

  const addProductToCart = (uid, data) => {
    setShoppingCart((prev) => {
      const index = prev.findIndex((el) => el.uid === uid);
      if (index >= 0) {
        return prev.map((el) =>
          el.uid === uid ? { ...el, quantity: Math.min(el.quantity++, 99) } : el
        );
      } else {
        data.quantity = 1;
        return [...prev, data];
      }
    });
  };

  const eventsCart = {
    removeItem: handleRemoveClick,
    inputQuantity: handleQuantityChange,
    increaseQuantity: handleIncreaseQuantity,
    decreaseQuantity: handleDecreaseQuantity,
    checkCart: handleCheckCart,
    emptyCart: handleEmptyCart,
    closeCart: handleCloseCart,
  };

  //////////////////////////////////////
  ///Catalogue Callbacks////////////////
  //////////////////////////////////////

  const handleOpenCart = () => setCartState({ display: true });

  const cartCount = () =>
    shoppingCart.reduce((acc, obj) => acc + obj.quantity, 0);

  const handleItemClick = (event) => {
    const uid = event.target.parentElement.id;
    const data = shoppingCatalogue.products.find((el) => el.uid === uid);
    addProductToCart(uid, data);
  };

  const eventsCatalogue = {
    addItem: handleItemClick,
    openCart: handleOpenCart,
   };

  //////////////////////////////////////
  ///Return (render) ///////////////////
  //////////////////////////////////////

  return (
    <div id="mainShop">
      <Catalogue
        products={shoppingCatalogue.products}
        events={eventsCatalogue}
        state={cartState}
        count={cartCount()}
      />
      <Cart 
      itemList={shoppingCart} 
      events={eventsCart}
      state={cartState}
      />
    </div>
  );
}

export default Main