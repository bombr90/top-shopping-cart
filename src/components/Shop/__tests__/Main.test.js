// __tests__/main.test.js

import React from "react";
import ReactDOM from 'react-dom'
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";  // optional
import userEvent from "@testing-library/user-event";
import Main from "./../Main";


describe('Smoke test of Main component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    render(<Main />, div)
  })
})

// localStorage mock
const localStorageMock = {
  getitem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}

// Component initialization and rendering testing
describe("Checking initialization and rendering of component", () => {
  it("renders correct headings", () => {
    const { getByRole } = render(<Main />)
    expect(screen.getByRole('heading',{name: 'Catalogue'}).textContent).toMatch('Catalogue')
    expect(screen.getByRole('heading',{name: 'Cart'}).textContent).toMatch('Cart')
    expect(getByRole).toMatchSnapshot();
    // Check that catalogue of 10 items has been rendered
    expect(screen.getAllByRole("button",{name: 'Add to Cart'}).length).toBe(10) 
  })
})

// Cart event function testing

function setup() {return {}}

describe('All Cart events/callbacks from this file', () => {
  describe("handleRemoveClick", () => {
    it("removes item from cart", () => {
    });
  });

  describe("handleQuantityChange", () => {
    it("changes item quantity in cart", () => {});
  });

  describe("handleIncreaseQuantity", () => {
    it("increases item quantity by 1", () => {});
  });

  describe("handleDecreaseQuantity", () => {
    it("deccreases item quantity by 1", () => {});
  });

  describe("handleCheckCart", () => {
    it("chceks out the cart", () => {});
  });

  describe("handleEmptyCart", () => {
    it("clears the cart after confirmation", () => {});
  });

  describe("handleCloseCart", () => {
    it("minimizes (hides) the cart component", () => {
      render(<Main />);
      const openButton = document.getElementById("cartIcon");
      const element = document.getElementById("cart");
      const button = document.getElementsByClassName("escape")[0];
      userEvent.click(openButton)
      expect(element).toBeVisible;
      userEvent.click(button);
      expect(element).not.toBeVisible;
    });
  });


})

//catalogue event function testing
describe("All Catalogue events/callbacks from this file", () => {
  describe("handleItemClick", () => {
    it("adds item from cart", () => {
    });
  });

  describe("handleOpenCart", () => {
    it("displays (un-hides) the cart component", () => {
      render(<Main />)
      const button = document.getElementById('cartIcon')
      const element = document.getElementById('cart')
      expect(element).not.toBeVisible;
      userEvent.click(button);
      expect(element).toBeVisible
    });
  });
});

//Internal function testing
describe("All local functions from this file", () => {
  describe("createProductsList", () => {
    it("generates a sample array of objects from imported data and appends a uid", () => {});
  });
  
  describe("cartCount", () => {



    it('Test Mock Component', () => {
      render(<Main />)
    })

    it("calculates total number of units in the cart", () => {
      render(<Main />)
      
    });
  });

});
