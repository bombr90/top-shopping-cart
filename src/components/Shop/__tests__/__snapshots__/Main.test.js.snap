// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checking initialization and rendering of component renders correct headings 1`] = `[Function]`;
